File: engine/src/Hermes.Engine/Domain/Entities/Monitoring.cs
namespace Hermes.Engine.Domain.Entities;

public class PipelineActivation : BaseEntity
{
    public Guid PipelineInstanceId { get; set; }
    public ActivationStatus Status { get; set; } = ActivationStatus.Starting;
    public DateTimeOffset StartedAt { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset? StoppedAt { get; set; }
    public DateTimeOffset? LastHeartbeatAt { get; set; }
    public DateTimeOffset? LastPolledAt { get; set; }
    public string? ErrorMessage { get; set; }
    public string? WorkerId { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public PipelineInstance PipelineInstance { get; set; } = null!;
    public List<WorkItem> WorkItems { get; set; } = new();
}

File: backend/hermes/domain/models/monitoring.py
"""Monitoring layer models - 'What IS running' (activations)."""

import uuid
from datetime import datetime

from sqlalchemy import ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from vessel.domain.models.base import Base


class PipelineActivation(Base):
    __tablename__ = "pipeline_activations"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default=uuid.uuid4, server_default=func.gen_random_uuid()
    )
    pipeline_instance_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("pipeline_instances.id", ondelete="RESTRICT"), nullable=False
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="STARTING")
    started_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    stopped_at: Mapped[datetime | None] = mapped_column()
    last_heartbeat_at: Mapped[datetime | None] = mapped_column()
    last_polled_at: Mapped[datetime | None] = mapped_column()
    error_message: Mapped[str | None] = mapped_column(Text)
    worker_id: Mapped[str | None] = mapped_column(String(256))
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)

    pipeline_instance: Mapped["PipelineInstance"] = relationship(  # type: ignore[name-defined]
        back_populates="activations",
    )
    work_items: Mapped[list["WorkItem"]] = relationship(  # type: ignore[name-defined]
        back_populates="pipeline_activation",
    )
